### PR TITLE
Prac 15 tutor profile form frontend

### PR DESCRIPTION
**Pull request by Kremifany**
## responsive

_What this code should do_

## Test Steps

_Instructions on how to test the changes_

## UI/UX Outcome

_Attach a screenshot here_

### CODE REVIEW

File: src/components/SubjectMultiselects.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    education: string;
+    about: string;
+    experience: string;
+}
+
+const SubjectMultiselects: React.FC<SubjectMultiselectsProps> = (props) => {
+    const { value, options, onChange } = useMultiSelect({
+        value: ['Geometry'],
+        options: [
+            { label: 'math', value: 'math' },
```
Comment (oxangyal) [inline src/components/SubjectMultiselects.tsx]: Please be consistent in naming,  needs to be corrected

File: src/models/interfaces.ts [hunk 2]
```diff
@@ -3,10 +3,39 @@ interface StudentTutorInfo {
...
     name: string;
     grade: string | number;
     _id: string;
     tutorInfo?: StudentTutorInfo;
 }
+export interface Subject {
+    _id: string; // Subject ID from the database
+    name: string; // Subject name
+}
+export interface Tutor extends User {
```
Comment (oxangyal) [inline src/models/interfaces.ts]: please pull the last updates.  We already have working  interfaces Tutor

File: src/pages/TutorProfilePage.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                    />
+                </Box>
+                <Spacer />
+                <HStack direction="row" spacing={7} display="flex" mt="30px">
+                    <Button
+                        type="submit"
+                        style={{ ...buttonStyle, backgroundColor: '#F4CD76', color: 'black' }}
+                        flex="1"
+                    >
+                        Submit
```
Comment (oxangyal) [inline src/pages/TutorProfilePage.tsx]: There is a different button in figma design, please add space to separate buttons from from

File: src/components/ProfileForTutor.tsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            <VStack>
+                <Textarea
+                    textColor="black"
+                    placeholder="About"
+                    h="200px"
+                    size="lg"
+                    bgColor="white"
+                />
+                <Spacer />
+                <Input textColor="black.400" placeholder="Experience" size="lg" bgColor="white" />
```
Comment (oxangyal) [inline src/components/ProfileForTutor.tsx]: Experience is drop down list not an input

File: src/components/ProfileForTutor.tsx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                        <AvatarBadge
+                            sx={{
+                                // paddingLeft: 8,
+                                marginBottom: 2,
+                            }}
+                            boxSize="0.9em"
+                        >
+                            <AddIcon
+                                sx={{
+                                    color: '#090759',
```
Comment (oxangyal) [inline src/components/ProfileForTutor.tsx]: We dont have this color in our project #090759,  the icon is black by default

File: src/components/ProfileForTutor.tsx (hunk 4 above)
Comment (Kremifany) [reply src/components/ProfileForTutor.tsx]: what are the options for that dropdown?

File: package.json [hunk 6]
```diff
@@ -17,9 +17,12 @@
   "dependencies": {
     "@chakra-ui/icons": "^2.1.1",
     "@chakra-ui/react": "^2.8.2",
+    "@cloudinary/react": "^1.13.0",
```
Comment (IrynaKolh) [inline package.json]: Uninstall these packages, we dont need this

File: src/pages/TutorProfilePage.tsx [hunk 7]
```diff
@@ -0,0 +1,766 @@
...
+
+const TutorProfilePage: React.FC<TutorProfilePageProps> = (isUpdate) => {
+    const { firstName, lastName, email } = JSON.parse(localStorage.getItem('userData') ?? '');
+    const [selectedImage, setSelectedImage] = useState<Blob | null>(null);
+    const [initialValues, setInitialValues] = useState<TutorRequest>({
+        grades: [],
+        about: '',
+        education: '',
+        avatar: '',
+        availability: ['Monday', 'Blanday'],
```
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx]: delete unexisted day of the week

File: src/pages/TutorProfilePage.tsx [hunk 8]
```diff
@@ -0,0 +1,766 @@
...
+    isUpdate: boolean;
+}
+import { MultiSelect, Option, useMultiSelect } from 'chakra-multiselect';
+import { AddIcon } from '@chakra-ui/icons';
+import avatar from '../assets/avatar.jpg';
+import { TutorRequest } from '../models/interfaces.ts';
+import axios from 'axios';
+import { theme } from '../util/theme.ts';
+import { headers } from '../util';
+//import { Cloudinary } from '@cloudinary/url-gen';
```
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx]: delete line

File: src/pages/TutorProfilePage.tsx [hunk 9]
```diff
@@ -0,0 +1,766 @@
...
+                            ForeignLanguages: values.ForeignLanguages,
+                            English: values.English,
+                            SocialStudies: values.SocialStudies,
+                            Science: values.Science,
+                        },
+                        { headers }
+                    );
+                    const { data, status } = response;
+                    console.log(data);
+                    if (status === 201) console.log('Tutor was updatedted successfully');
```
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx]: Typo error "updated"

File: src/pages/TutorProfilePage.tsx [hunk 10]
```diff
@@ -0,0 +1,766 @@
...
+            'ACT Science Test Prep',
+        ],
+        _id: '',
+        userId: '',
+        __v: 0,
+    };
+
+    //FETCHING TUTOR DATA FOR AN UPDATE
+    useEffect(() => {
+        if (!isUpdate) {
```
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx]: What sense of this code?
 if (!isUpdate) {
            return;
        }

File: src/pages/TutorProfilePage.tsx [hunk 11]
```diff
@@ -0,0 +1,766 @@
...
+            return;
+        }
+        const fetchTutor = async () => {
+            try {
+                const tutorId = localStorage.getItem('tutorId');
+                const response = await axios.get(
+                    `https://ffprac-team2-back.onrender.com/api/v1/tutors/${tutorId}`
```
Comment (oxangyal) [inline src/pages/TutorProfilePage.tsx]: For best practices I would recommend to hide the full URL used for making requests in the .env file. This guarantees a more secure management of confidential data such as API URLs and helps the process of changing URLs without modifying the application code. I would use something like REACT_APP_BASE_URL for instance.

File: src/App.tsx [hunk 12]
```diff
@@ -25,6 +28,7 @@ const App = () => {
                     <Route path="/auth" element={<AuthPage activeTab={'login'} />} />
                     {/* <Route path="/login" element={<LoginForm />} /> */}
                     <Route path="/parent-dashboard" element={<ParentDashboardPage />} />
+                    <Route path="/tutor-profile" element={<TutorProfilePage isUpdate={true} />} />
```
Comment (IrynaKolh) [inline src/App.tsx]: You don't need to pass props. Use useState hook in component.

File: src/pages/TutorProfilePage.tsx [hunk 13]
```diff
@@ -0,0 +1,766 @@
...
+
+                                    {/* Experience*/}
+                                    <FormControl>
+                                        <Field name="yearsOfExperience">
+                                            {({ field, form }) => (
+                                                <MultiSelect
+                                                    {...field}
+                                                    options={yearsOfExperienceOptions}
+                                                    value={form.values.yearsOfExperience} //for displaying selected values
+                                                    label="yearsOfExperience"
```
Comment (oxangyal) [inline src/pages/TutorProfilePage.tsx]: pls change label to "Experience (in years)"

File: src/validationSchemas/index.ts [hunk 14]
```diff
@@ -54,6 +54,81 @@ export const connectSchema = yup.object().shape({
...
+            .oneOf([
+                'Biology',
+                'Chemistry',
+                'Earth Science',
+                'Physics',
+                'Science',
+                'ACT Science Test Prep',
+            ])
+    ),
+    _id: yup.string(),
```
Comment (oxangyal) [inline src/validationSchemas/index.ts:128]: We are not validating these fields:   _id: yup.string(),
    userId: yup.string(),
    __v: yup.number(), please remove them

File: src/models/interfaces.ts [hunk 15]
```diff
@@ -4,7 +4,20 @@ interface StudentTutorInfo {
     availability: string;
     tutorName: string;
 }
+export interface User {
```
Comment (IrynaKolh) [inline src/models/interfaces.ts]: You don't use this interface in code. Delete it.

File: src/models/interfaces.ts [hunk 16]
```diff
@@ -4,7 +4,20 @@ interface StudentTutorInfo {
...
+    user_id: string; // Auto-generated
+    firstName: string; // First name with required validation
+    lastName: string; // Last name with required validation
+    email: string; // Email with required validation
+    password: string; // Password with required validation
+    role: UserRole; // Role using a UserRole enum
+}
 
+export enum UserRole {
```
Comment (IrynaKolh) [inline src/models/interfaces.ts]: You don't use this enum in code. Delete it.

File: src/models/interfaces.ts [hunk 17]
```diff
@@ -69,6 +82,51 @@ export interface GlobalState {
     // teacherSearchResults: string[];
     updateStudents: () => Promise<void>;
 }
+export interface Subject {
+    _id: string; // Subject ID from the database
+    name: string; // Subject name
+}
+export interface Tutor {
```
Comment (IrynaKolh) [inline src/models/interfaces.ts]: We already have the Tutor interface in line 50, edit name (forExample TutorById)

Your interface must be equal response from server and looks like this:

export interface TutorById {
    availability: string[]; //availability is array : needed change for multiselect in tutorProfilePage
    userId: string;
    about: string;
    grades: string[];
    avatar?: string;
    education: string;
    yearsOfExperience: number;
    MathSubject: string[];
    ForeignLanguages: string[];
    English: string[];
    SocialStudies: string[];
    Science: string[];
    _id: string;
    _v: string;
}


In addition you don't use this interface anywhere

File: src/models/interfaces.ts [hunk 18]
```diff
@@ -69,6 +82,51 @@ export interface GlobalState {
...
+    avatar?: string;
+    education: string;
+    yearsOfExperience: number;
+    MathSubject: string[];
+    ForeignLanguages: string[];
+    English: string[];
+    SocialStudies: string[];
+    Science: string[];
+    _id: string;
+    userId: string;
```
Comment (IrynaKolh) [inline src/models/interfaces.ts]: userId we pass through headers

File: src/models/interfaces.ts [hunk 19]
```diff
@@ -69,6 +82,51 @@ export interface GlobalState {
     // teacherSearchResults: string[];
     updateStudents: () => Promise<void>;
 }
+export interface Subject {
```
Comment (IrynaKolh) [inline src/models/interfaces.ts]: You don't use this Subject interface  in code. Delete it.

File: src/models/interfaces.ts [hunk 20]
```diff
@@ -69,6 +82,51 @@ export interface GlobalState {
...
+    MathSubject: string[];
+    ForeignLanguages: string[];
+    English: string[];
+    SocialStudies: string[];
+    Science: string[];
+    _id: string;
+    userId: string;
+    __v: number;
+}
+export interface AuthPageProps {
```
Comment (IrynaKolh) [inline src/models/interfaces.ts:86]: Delete duplicate (Alena created this )

File: src/pages/TutorProfilePage.tsx [hunk 21]
```diff
@@ -0,0 +1,766 @@
...
+            })
+                .then((response) => {
+                    return response.text();
+                })
+                .then((data) => {
+                    // document.getElementById('data').innerHTML += data;
+                    console.log('Upload success', data);
+                });
+        }
+        if (isUpdate) {
```
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx]: use useState to handle this condition

File: src/pages/TutorProfilePage.tsx [hunk 22]
```diff
@@ -0,0 +1,766 @@
...
+                                                                    (item) => item.value
+                                                                )
+                                                            ); //updates selected options in tutorRequest
+                                                        }}
+                                                    />
+                                                )}
+                                            </Field>
+                                        </FormControl>
+                                    </Box>
+                                    {/* GRADES*/}
```
Comment (oxangyal) [inline src/pages/TutorProfilePage.tsx]: Exchange  fields:  "Grades" and "Foreign languages" for better flow

File: src/models/interfaces.ts [hunk 23]
```diff
@@ -69,6 +82,51 @@ export interface GlobalState {
...
+    grades: string[];
+    avatar?: string;
+    education: string;
+    yearsOfExperience: number;
+    MathSubject: string[];
+    ForeignLanguages: string[];
+    English: string[];
+    SocialStudies: string[];
+    Science: string[];
+    _id: string;
```
Comment (IrynaKolh) [inline src/models/interfaces.ts]: _id you pass through url, remove this line

File: src/validationSchemas/index.ts [hunk 24]
```diff
@@ -54,6 +54,81 @@ export const connectSchema = yup.object().shape({
...
+    grades: yup.array(
+        yup.string().oneOf(['K', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'])
+    ),
+    availability: yup.array(
+        yup
+            .string()
+            .oneOf(['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'])
+    ),
+    education: yup.string(),
+    yearsOfExperience: yup.number(),
```
Comment (IrynaKolh) [inline src/validationSchemas/index.ts]: yearsOfExperience: {
        type: Number,
        min: 1,
        max: 50,
    },

File: src/pages/TutorProfilePage.tsx [hunk 25]
```diff
@@ -0,0 +1,766 @@
...
+            'Thursday',
+            'Friday',
+            'Saturday',
+            'Sunday',
+        ],
+        about: '',
+        grades: ['K', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'],
+        avatar: '',
+        education: '',
+        yearsOfExperience: 0,
```
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx]: yearsOfExperience: {
        type: Number,
        min: 1,
        max: 50,
    },
    
    min 1

File: src/pages/TutorProfilePage.tsx (hunk 9 above)
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx]: status = 200

File: src/pages/TutorProfilePage.tsx [hunk 26]
```diff
@@ -0,0 +1,766 @@
...
+                            English: values.English,
+                            SocialStudies: values.SocialStudies,
+                            Science: values.Science,
+                        },
+                        { headers }
+                    );
+                    const { data, status } = response;
+                    console.log(data);
+                    if (status === 201) console.log('Tutor was updatedted successfully');
+                    if (status !== 201) {
```
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx]: status !=200

File: src/pages/TutorProfilePage.tsx [hunk 27]
```diff
@@ -0,0 +1,766 @@
...
+                                        Submit
+                                    </Button>
+                                    <Button
+                                        type="button"
+                                        variant="buttonTeal"
+                                        size="lg"
+                                        fontWeight="bold"
+                                        width="175px"
+                                    >
+                                        Cancel
```
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx:761]: On Cancel nothing happend

**Review by IrynaKolh — COMMENTED** [review]
Very nice code, but has some small bugs, fix them and I will check again)

**Comment by Kremifany** [comment]
PR for tutor profile form page

_What code should do:_
Implements Tutor profile page functionality with initial Tutor profile creation with possibility to update profile
Test Steps:
1.**Create a Tutor** - IsUpdate in page props in App.tsx. should be false. Submit button create a Tutor.
1. **Update a Tutor**:
                              - register a Tutor and Create a Tutor, only then you can proceed to update a Tutor .
                              - isUpdate  in page props should be true  in page props tutorProfilePage route in app.tsx,
                              - reload the page
                              - make sure that fields are filled with data from created or updated Tutor.
                              - Submit button updates the latest info.
                               TBD navigation later isUpdate will work from NavBar button.
2.**Upload Image**:
                              - from computer works by using Plus button on avatar and choosing an image file.
                              - avatar not saved yet, waiting for API
3.
UI/UX Outcome
image

File: src/pages/TutorProfilePage.tsx [hunk 28]
```diff
@@ -0,0 +1,803 @@
...
+                                    mt="30px"
+                                >
+                                    <Button
+                                        type="submit"
+                                        variant="buttonYellow"
+                                        size="lg"
+                                        fontWeight="bold"
+                                        width="175px"
+                                    >
+                                        Submit
```
Comment (oxangyal) [inline src/pages/TutorProfilePage.tsx]: Button "Save"

File: src/App.tsx (hunk 12 above)
Comment (Kremifany) [reply src/App.tsx]: Please clarify where from this state will come  to this component

File: src/models/interfaces.ts (hunk 17 above)
Comment (Kremifany) [reply src/models/interfaces.ts]: removed extra tutor interface caused by merge by mistake

File: src/pages/TutorProfilePage.tsx (hunk 21 above)
Comment (Kremifany) [reply src/pages/TutorProfilePage.tsx]: I am changing the behavior of component through the props that will be updated from navbar, as discussed.

File: src/App.tsx [hunk 29]
```diff
@@ -25,6 +28,7 @@ const App = () => {
                     <Route path="/auth" element={<AuthPage activeTab={'login'} />} />
                     {/* <Route path="/login" element={<LoginForm />} /> */}
                     <Route path="/parent-dashboard" element={<ParentDashboardPage />} />
+                    <Route path="/tutor-profile" element={<TutorProfilePage isUpdate />} />
```
Comment (IrynaKolh) [inline src/App.tsx]: isUpdate  delete

File: src/pages/TutorProfilePage.tsx [hunk 30]
```diff
@@ -0,0 +1,792 @@
...
+}
+import { MultiSelect, Option, useMultiSelect } from 'chakra-multiselect';
+import { AddIcon } from '@chakra-ui/icons';
+import avatar from '../assets/avatar.jpg';
+import { TutorRequest } from '../models/interfaces.ts';
+import axios from 'axios';
+import { theme } from '../util/theme.ts';
+import { headers } from '../util';
+
+const TutorProfilePage: React.FC<TutorProfilePageProps> = ({ isUpdate = false }) => {
```
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx]: delete { isUpdate = false }

File: src/pages/TutorProfilePage.tsx [hunk 31]
```diff
@@ -0,0 +1,792 @@
...
+    const { firstName, lastName, email } = JSON.parse(localStorage.getItem('userData') ?? '{}');
```
Comment (IrynaKolh) [inline src/pages/TutorProfilePage.tsx:32]: add 1 line 
const [isUpdate, setIsUpdate] = useState(true);

then change isUpdate accorting your logic

**Review by IrynaKolh — COMMENTED** [review]
Looks very good) 
use in terminal 'npm run format' for fixing prettier and linter errors, because checks didn't pass

File: src/pages/TutorProfilePage.tsx (hunk 13 above)
Comment (Kremifany) [reply src/pages/TutorProfilePage.tsx]: Done!

File: src/pages/TutorProfilePage.tsx (hunk 22 above)
Comment (Kremifany) [reply src/pages/TutorProfilePage.tsx]: Yep!

File: src/App.tsx (hunk 29 above)
Comment (Kremifany) [reply src/App.tsx]: isUpdate flag is used in order to differentiate between update and create a new tutor.

File: src/pages/TutorProfilePage.tsx (hunk 31 above)
Comment (Kremifany) [reply src/pages/TutorProfilePage.tsx:32]: You suggest to use this state inside the function component, but where is that state will take his value to set? it should come to a component from outside with props.  
What is the user flow that will change the state you suggest to use within  the component?

**Comment by Kremifany** [comment]
Minor change in a props to a component: 
-Passing isUpdate through props means that component will function as update tutor 
-default is FALSE, so it creates Tutor

File: src/pages/TutorProfilePage.tsx (hunk 27 above)
Comment (Kremifany) [reply src/pages/TutorProfilePage.tsx:761]: what is expected?

Comment (Kremifany) [reply src/pages/TutorProfilePage.tsx:761]: 1.on update
2. on create

Comment (IrynaKolh) [reply src/pages/TutorProfilePage.tsx:761]: I expected to close form

Comment (oxangyal) [reply src/pages/TutorProfilePage.tsx:761]: I think  tutor should be directed to the tutordashboard?

File: src/pages/TutorProfilePage.tsx (hunk 31 above)
Comment (IrynaKolh) [reply src/pages/TutorProfilePage.tsx:32]: you pass the same props all time. If it has never changed the best way to use in component useState hook and
1) if you get tutor diring fetching data set to true/false (according to your logic)
2) in createTutor function set to true/false  (according to your logic)

Comment (Kremifany) [reply src/pages/TutorProfilePage.tsx:32]: you pass the same props all time. - not true,
isUpdate prop can be true and false depends on logic from outside the component 
When isUpdate passed in props - means it's true and then update tutor logic works within the component
When not passed - the default value is used : isUpdate=false and then its a Create tutor.

Comment (IrynaKolh) [reply src/pages/TutorProfilePage.tsx:32]: When and who pass isUpdate = true??? Show me in a code

**Review by IrynaKolh — APPROVED** [review]
Greate Job! LGFM

**Review by mynewyear — APPROVED** [review]
LGTM

**Pull request merged**